### PR TITLE
[DNM] Doubly linked skiplist by Wei Liu

### PR DESCRIPTION
**Pull request by yiwu-arbug**
The patch is done by @Little-Wallace. Just send it here for code review purpose.

### CODE REVIEW

File: memtable/doubly_linked_skiplist.h [hunk 1]
```diff
@@ -0,0 +1,1002 @@
...
+               compare_(x->Key(), splice->next_[i]->Key()) < 0);
+        assert(splice->prev_[i] == head_ ||
+               compare_(splice->prev_[i]->Key(), x->Key()) < 0);
+        x->NoBarrier_SetNext(i, splice->next_[i]);
+        if (i == 0) {
+            x->NoBarrier_SetPrev(splice->prev_[0]);
+        }
+        if (splice->prev_[i]->CASNext(i, splice->next_[i], x)) {
+          // success
+          if (UNLIKELY(i == 0 && splice->next_[0] != nullptr)) {
```
Comment (yiwu-arbug) [inline memtable/doubly_linked_skiplist.h]: Not `UNLIKELY` at all.

Comment (yiwu-arbug) [inline memtable/doubly_linked_skiplist.h]: IMO `CASPrev` should be done independent with `CASNext`. For example, here if `CASNext` success but `CASPrev` fail, we should only retry `CASPrev`.

**Comment by yiwu-arbug** [comment]
@Little-Wallace Here's my comment based on your branch. After you update your code, you can send another PR to pingcap/rocksdb for code review, and I'll close this one. Thanks.

Also cc @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
Don't merge to master but to a branch.

**Comment by yiwu-arbug** [comment]
Closing in favor of #103

**Pull request closed without merging**